### PR TITLE
Directory structure and manifest parsing

### PR DESCRIPTION
**Pull request by jasongin**
Most of the library code is unaffected by the directory structure and manifest format changes, because those things are abstracted away by the `PackageSource` and `PackageInfo` classes. There are 0 changes to public APIs.

Almost all the files in the change list here are updates to test schemas and translators and test cases, which are updated to match the latest schema and translator formats we've discussed. Significant non-test changes are only in 3 files:

**`InstalledPackageSource`** - This is renamed from `LocalPackageSource`, and now inherits from the new class of that name (so it can share some protected helper methods). As before, this class supports loading info from packages installed to the application's "node_modules" directory.
**`LocalPackageSource`** - This is rewritten to support loading package information from a locally-cloned git source directory, even when the source files have not been modified by any npm-packaging build script. All the code for dealing with the directory structure is here, and while it's not pretty at least it is confined.
**`PackageSource`** - The createLocalPackageSource static method now creates either a `LocalPackageSource` or `InstalledPackageSource` depending on whether the specified path points to a "node_modules" directory.

Not included in this change:
- OCF schema reader
- There will be a build script that will use some of the methods in LocalPackageSource to gather information about packages in order to create NPM packages with the correct files and metadata.
- There are still some unresolved TODOs related to onboarding that can be filled in later.

### CODE REVIEW

File: node/test/@opent2t/test-b/TranslatorBC.js [hunk 1]
```diff
@@ -1,78 +0,0 @@
...
-const EventEmitter = require("events");
-
-/**
- * This translator class implements two interfaces: "org.opent2t.test.B"
- * and "org.opent2t.test.C". Since those interfaces do not have any conflicting
- * member names, they are both implemented directly.
- */
-class TestTranslatorBC extends EventEmitter {
-
-    constructor(deviceProps) {
```
Comment (tjaffri) [inline node/test/@opent2t/test-b/TranslatorBC.js:14]: I think I recall we decided to stick with initDevice pattern for now. That's what I have in my PR, at least.

EDIT: Thought some more about this and constructor is way better. I think the argument was that constructors are "too complicated" and we can explain "initiDevice" better... but I don't think that argument is valid. If you are able to write hello world, you know what a constructor is in pretty much any language.

So I updated my PR with the constructor pattern as well. We can discuss further if others disagree, but I think you and I agree on this.

File: node/lib/package/InstalledPackageSource.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    public async getAllPackageInfoAsync(): Promise<PackageInfo[]> {
+        let packageInfos: PackageInfo[] = [];
+        let directoryNames: string[] =
+                await LocalPackageSource.getSubdirectoriesAsync(this.cacheDirectory);
+
+        for (let i = 0; i < directoryNames.length; i++) {
+            let packageName: string = directoryNames[i];
+            if (packageName.startsWith("@")) {
+                let subdirectoryNames: string[] = await LocalPackageSource.getSubdirectoriesAsync(
+                        path.join(this.cacheDirectory, packageName));
```
Comment (sanjaiganesh) [inline node/lib/package/InstalledPackageSource.ts:47]: Hope there is no concept of 'nested' scopes. #ByDesign

File: node/test/@opent2t/test-b/TranslatorBC.js (hunk 1 above)
Comment (jasongin) [inline node/test/@opent2t/test-b/TranslatorBC.js:14]: Note this code is red, meaning it was deleted. :)

But I didn't switch to initDevice() in the new test translator code. That's orthogonal to the rest of these changes anyway, so we can change that later, or not.

File: node/lib/package/InstalledPackageSource.ts (hunk 2 above)
Comment (jasongin) [inline node/lib/package/InstalledPackageSource.ts:47]: No, there is no such thing as nested scopes. #ByDesign

File: node/lib/package/LocalPackageSource.ts [hunk 3]
```diff
@@ -90,47 +155,228 @@ export class LocalPackageSource extends PackageSource {
      * package, or null if the requested package is not found at the source
      */
     public async getPackageInfoAsync(name: string): Promise<PackageInfo | null> {
-        let packageJsonPath: string = path.join(this.cacheDirectory, name, "package.json");
+        if (name.startsWith("opent2t-schema-")) {
```
Comment (sanjaiganesh) [inline node/lib/package/LocalPackageSource.ts:158]: > opent2t-schema [](start = 29, length = 14)

Probably having "opent2t-schema", "opent2t-translator" etc.. as global defines will help.

**Comment by sanjaiganesh** [comment]
:shipit:

File: node/test/@opent2t/test-b/TranslatorBC.js (hunk 1 above)
Comment (sanjaiganesh) [inline node/test/@opent2t/test-b/TranslatorBC.js:14]: @tjaffri If you agree, I shall merge.

---

In reply to: [75169591](https://github.com/openT2T/opent2t/pull/20#discussion_r75169591) [](ancestors = 75169591)

File: node/lib/package/LocalPackageSource.ts (hunk 3 above)
Comment (jasongin) [inline node/lib/package/LocalPackageSource.ts:158]: Sure, I'll clean that up tonight. But you can go ahead and merge without waiting for this fix if you want.

File: node/test/@opent2t/test-b/TranslatorBC.js (hunk 1 above)
Comment (tjaffri) [inline node/test/@opent2t/test-b/TranslatorBC.js:14]: Thanks... yup I understand this is deleted code it just jogged my memory and I was intending to say let's keep constructor pattern... makes my code (in my PR) easier to read.

@sanjaiganesh :shipit: please merge.

**Comment by tjaffri** [comment]
:shipit:

**Pull request merged**